### ISSUE TITLE
Negative samples during training

### ISSUE THREAD

**Issue by adriaciurana** [dbolya/yolact#137]
First of all I want to thank the work done with YolAct.

You see, I'm trying to get the model to admit images that have NO objects. This is very useful for me because these images without objects help the model a lot to know the differences if the background and foreground are similar.

I know that some changes have to be made to the MultiboxLoss but at the moment I'm going something lost, could you give me some indications?

Thank you

**Comment by dbolya** [comment]
Someone was trying to do something similar and we came up with a couple of hacks to make it work (#89). I think you should probably start with that, because if you want to explicitly train on images without objects, you'd need to basically check if the annotations tensor was empty every time you used it which would be a pain.

**Comment by abhigoku10** [comment]
@adriaciurana  are you looking into the concept of negative hardmining or postivie hard mining ???
@dbolya the reference link you shared is for ignoring if objects are not annotated in a an image but what if i want to do something like hard mining ? can i use the approach you have mentioned

**Comment by dbolya** [comment]
@abhigoku10 Uhh this person is asking to train on images without annotations, which you can do by adding a dummy annotation to each image.

Hard mining is something completely different where you selectively train on the hardest examples (the ones that give the most loss). Yolact already does hard negative mining though.

**Comment by adriaciurana** [comment]
Thank you very much for the feedback. No, I do not want to do negative mining in fact from what I have seen yolact uses an SSD approach.

The images I work with are damage to industrial parts with a very specific image, for that reason a defective part and another one that is fine are very similar.

The problem of taking samples randomly, is that although it is true that many parts of the background are taken there are practically no good samples of parts without defects. What ends up happening is that it relates the defective being with the piece itself (since it has not seen enough).

That is why I want to give feedback to the system so that it sees completely clean images, without objects.

I will try to make the necessary changes in the MultiBox_Loss, I think I should only add the check if the target is empty and avoid the match. Line 105

And then add checks on those elements that depend on the anchors (localization and mask). Probably return a 0 directly?

**Comment by dbolya** [comment]
The issue is that the whole pipeline (not just the loss computation) assumes that the image has at least one annotation. In fact, the dataset (`data/coco.py`) won't even consider the image if it doesn't have any annotations. That's why I recommended to add a dummy annotation instead, so there would be at least one annotation in each image that you can just discard later.

**Comment by adriaciurana** [comment]
> The issue is that the whole pipeline (not just the loss computation) assumes that the image has at least one annotation. In fact, the dataset (`data/coco.py`) won't even consider the image if it doesn't have any annotations. That's why I recommended to add a dummy annotation instead, so there would be at least one annotation in each image that you can just discard later.

I'm currently using a custom dataset, and I don't evaluate with the usual metrics either.

Something fast that can be done to make it compatible with the entire pipeline is to train with a dataset that admits no objects and evaluate with one that only returns images with objects (perhaps a simple basary flag).

**Comment by dbolya** [comment]
Sorry, when I said "dataset", I meant the class I wrote to load the data, not necessarily COCO (I pointed to `data/coco.py` to say where it was implemented if you wanted to try to change it).

If you can get the whole pipeline to accept empty images, then that would definitely work, but do know that you should start with the dataloader and augmentations back at the beginning of the pipeline (or the images won't be loaded at all).

**Comment by abhigoku10** [comment]
@dbolya  "Yolact already does hard negative mining though." can you please elaborate a bit on this , how and where this functionality is implemented in the code

**Comment by dbolya** [comment]
@abhigoku10 We use OHEM just like SSD does: when training the class confidence, since we have so many anchors there's like 100x the negatives as positives. Since training with all the negatives wouldn't be good, we only train with a subset of the negatives (and all of the positives). So we choose `3 * num_positives` of the hardest negatives (the negatives with the highest loss) to train on instead of all of them.

It's implemented on this line:
https://github.com/dbolya/yolact/blob/605b3c8fdf283642e7753c355bc6ee02f8db1380/layers/modules/multibox_loss.py#L245

Read the OHEM paper for more details.

**Comment by dbolya** [comment]
Since this has been open so long, I'm going to close it. Feel free to reopen if you have any updates.

**Comment by PrajwalCogniac** [comment]
Hey @adriaciurana
I am also working on images where I have to do defect detection and facing exactly same problem as you where object with interest has very similar features with object of no interest. So did you make changes to Multiloss function or how did you go about handling it ? 
Any help on this issue is useful @adriaciurana @dbolya @abhigoku10
